accessor.go

```diff
@@ -54,11 +54,34 @@ func (sa *ShardAccessor) Shard() shard.Key {
 	return sa.shard.key
 }
 
+func (sa *ShardAccessor) Read(p []byte) (int, error) {
+	sa.lk.Lock()
+	defer sa.lk.Unlock()
+
+	if sa.mmapr == nil {
+		if f, ok := sa.data.(*os.File); ok {
+			if mmapr, err := mmap.Open(f.Name()); err != nil {
+				log.Warnf("failed to mmap reader of type %T: %s; using reader as-is", sa.data, err)
+			} else {
+				sa.mmapr = mmapr
+			}
+		}
+	}
+
+	if sa.mmapr != nil {
+		return sa.mmapr.ReadAt(p, 0)
+	}
+
+	return sa.data.Read(p)
+}
+
 func (sa *ShardAccessor) Blockstore() (ReadBlockstore, error) {
 	var r io.ReaderAt = sa.data
 
 	sa.lk.Lock()
-	if f, ok := sa.data.(*os.File); ok {
+	if sa.mmapr != nil {
+		r = sa.mmapr
+	} else if f, ok := sa.data.(*os.File); ok {
```

Comment:
Let's extract the mmap logic; then make this method short-circuit for more readable and idiomatic code.